File: packages/react-native-fipa/android/src/main/java/com/deviceattestation/DeviceAttestationSessionVault.kt
package com.deviceattestation

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext

internal class DeviceAttestationSessionVault(context: ReactApplicationContext) :
    NativeSessionVaultSpec(context) {
    private val vault = SessionVault(AndroidVaultStorage.create(context))
    private val operations =
        BridgeOperations(
            1,
            64,
            { error ->
                (error as? NativeFailure)?.code?.takeIf { it.startsWith("vault_") }
                    ?: "vault_storage_failed"
            },
            "Unable to complete secure session storage operation.",
            NativeFailure("vault_unavailable"),
        )

    override fun acquire(
        storageNamespace: String,
        slotId: String,
        accessibility: String,
        leaseMilliseconds: Double,
        preserveSession: Boolean,
        promise: Promise,
    ) =
        operations.execute(promise) {
            val snapshot =
                vault.acquire(
                    storageNamespace,
                    slotId,
                    accessibility,
                    leaseMilliseconds,
                    preserveSession,
                )
            Arguments.createMap().apply {
                putString("leaseId", snapshot.leaseId)
                putDouble("generation", snapshot.generation.toDouble())
                putString("identityJSON", snapshot.identityJSON)
                putString("sessionJSON", snapshot.sessionJSON)
                putBoolean("recoveryRequired", snapshot.recoveryRequired)
            }
        }

    override fun commit(
        storageNamespace: String,
        slotId: String,
        leaseId: String,
        generation: Double,
        identityJSON: String?,
        sessionJSON: String?,
        recoverySessionJSON: String?,
        hasInteraction: Boolean,
        promise: Promise,
    ) =
        operations.execute(promise) {
            vault
                .commit(
                    storageNamespace,
                    slotId,
                    leaseId,
                    generation,
                    identityJSON,
                    sessionJSON,
                    recoverySessionJSON,
                    hasInteraction,
                )
                .toDouble()
        }

    override fun renew(
        storageNamespace: String,
        slotId: String,
        leaseId: String,
        generation: Double,
        leaseMilliseconds: Double,
        promise: Promise,
    ) =
        operations.execute(promise) {
            vault.renew(storageNamespace, slotId, leaseId, generation, leaseMilliseconds)
            null
        }

    override fun abandon(
        storageNamespace: String,
        slotId: String,
        leaseId: String,
        generation: Double,
        promise: Promise,
    ) =
        operations.execute(promise) {
            vault.abandon(storageNamespace, slotId, leaseId, generation).toDouble()
        }

    override fun invalidate(
        storageNamespace: String,
        slotId: String,
        accessibility: String,
        promise: Promise,
    ) =
        operations.execute(promise) {
            vault.invalidate(storageNamespace, slotId, accessibility).toDouble()
        }

    override fun discard(
        storageNamespace: String,
        slotId: String,
        generation: Double,
        promise: Promise,
    ) = operations.execute(promise) { vault.discard(storageNamespace, slotId, generation) }

    override fun saveIdentity(
        storageNamespace: String,
        slotId: String,
        leaseId: String,
        generation: Double,
        identityJSON: String,
        promise: Promise,
    ) =
        operations.execute(promise) {
            vault.saveIdentity(storageNamespace, slotId, leaseId, generation, identityJSON)
            null
        }

    override fun clearSession(
        storageNamespace: String,
        slotId: String,
        leaseId: String,
        generation: Double,
        promise: Promise,
    ) =
        operations.execute(promise) {
            vault.clearSession(storageNamespace, slotId, leaseId, generation)
            null
        }

    override fun release(
        storageNamespace: String,
        slotId: String,
        leaseId: String,
        generation: Double,
        promise: Promise,
    ) =
        operations.execute(promise) {
            vault.release(storageNamespace, slotId, leaseId, generation)
            null
        }

    override fun cancelInteraction(
        storageNamespace: String,
        slotId: String,
        accessibility: String,
        promise: Promise,
    ) =
        operations.execute(promise) {
            val result = vault.cancelInteraction(storageNamespace, slotId, accessibility)
            Arguments.createMap().apply {
                putString("sessionJSON", result.sessionJSON)
                putBoolean("cancelled", result.cancelled)
            }
        }

    override fun invalidate() {
        operations.invalidate()
        super.invalidate()
    }
}

File: packages/react-native-fipa/ios/DeviceAttestationSessionVault.swift
import Foundation
import React

@objc(DeviceAttestationSessionVault)
class DeviceAttestationSessionVault: NSObject {
  private static let queue = DispatchQueue(label: "DeviceAttestation.SessionVault")
  private static let vault: SessionVault? = {
    guard let bundle = Bundle.main.bundleIdentifier, !bundle.isEmpty else { return nil }
    return SessionVault(storage: KeychainSessionVaultStorage(service: "\(bundle).DeviceAttestation.SessionVault.v1"))
  }()

  @objc static func requiresMainQueueSetup() -> Bool { false }

  @objc func acquire(_ namespace: String, slotId: String, accessibility: String,
                     leaseMilliseconds: Double, preserveSession: Bool, resolver resolve: @escaping RCTPromiseResolveBlock,
                     rejecter reject: @escaping RCTPromiseRejectBlock) {
    run(resolve, reject) { vault in
      guard let policy = VaultAccessibility(rawValue: accessibility) else { throw SessionVaultError.invalidInput }
      let snapshot = try vault.acquire(namespace: namespace, slot: slotId, accessibility: policy, leaseMilliseconds: leaseMilliseconds, preserveSession: preserveSession)
      return ["leaseId": snapshot.leaseId, "generation": NSNumber(value: snapshot.generation),
              "identityJSON": snapshot.identityJSON as Any? ?? NSNull(),
              "sessionJSON": snapshot.sessionJSON as Any? ?? NSNull(),
              "recoveryRequired": snapshot.recoveryRequired]
    }
  }

  @objc func commit(_ namespace: String, slotId: String, leaseId: String, generation: Double,
                    identityJSON: String?, sessionJSON: String?, recoverySessionJSON: String?, hasInteraction: Bool, resolver resolve: @escaping RCTPromiseResolveBlock,
                    rejecter reject: @escaping RCTPromiseRejectBlock) {
    run(resolve, reject) { vault in
      NSNumber(value: try vault.commit(namespace: namespace, slot: slotId, leaseId: leaseId,
        generation: generation, identityJSON: identityJSON, sessionJSON: sessionJSON,
        recoverySessionJSON: recoverySessionJSON, hasInteraction: hasInteraction))
    }
  }

  @objc func renew(_ namespace: String, slotId: String, leaseId: String, generation: Double,
                   leaseMilliseconds: Double, resolver resolve: @escaping RCTPromiseResolveBlock,
                   rejecter reject: @escaping RCTPromiseRejectBlock) {
    run(resolve, reject) { vault in
      try vault.renew(namespace: namespace, slot: slotId, leaseId: leaseId, generation: generation,
                      leaseMilliseconds: leaseMilliseconds)
      return nil
    }
  }

  @objc func abandon(_ namespace: String, slotId: String, leaseId: String, generation: Double,
                     resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    run(resolve, reject) { vault in
      NSNumber(value: try vault.abandon(namespace: namespace, slot: slotId, leaseId: leaseId, generation: generation))
    }
  }

  @objc func invalidate(_ namespace: String, slotId: String, accessibility: String,
                        resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    run(resolve, reject) { vault in
      guard let policy = VaultAccessibility(rawValue: accessibility) else { throw SessionVaultError.invalidInput }
      return NSNumber(value: try vault.invalidate(namespace: namespace, slot: slotId, accessibility: policy))
    }
  }

  private func run(_ resolve: @escaping RCTPromiseResolveBlock, _ reject: @escaping RCTPromiseRejectBlock,
                   _ operation: @escaping (SessionVault) throws -> Any?) {
    Self.queue.async {
      do {
        guard let vault = Self.vault else { throw SessionVaultError.unavailable }
        resolve(try operation(vault))
      } catch let error as SessionVaultError {
        reject(error.rawValue, "Unable to complete secure session storage operation.", nil)
      } catch {
        reject(SessionVaultError.storage.rawValue, "Unable to complete secure session storage operation.", nil)
      }
    }
  }

  @objc func discard(_ storageNamespace: String, slotId: String, generation: Double,
                     resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    run(resolve, reject) { vault in
      try vault.discard(namespace: storageNamespace, slot: slotId, generation: generation)
    }
  }
  @objc func saveIdentity(_ storageNamespace: String, slotId: String, leaseId: String, generation: Double,
                         identityJSON: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    run(resolve, reject) { vault in
      try vault.saveIdentity(namespace: storageNamespace, slot: slotId, leaseId: leaseId,
        generation: generation, identityJSON: identityJSON)
      return nil
    }
  }
  @objc func clearSession(_ storageNamespace: String, slotId: String, leaseId: String, generation: Double,
                         resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    run(resolve, reject) { vault in
      try vault.clearSession(namespace: storageNamespace, slot: slotId, leaseId: leaseId, generation: generation)
      return nil
    }
  }
  @objc func release(_ storageNamespace: String, slotId: String, leaseId: String, generation: Double,
                    resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    run(resolve, reject) { vault in
      try vault.release(namespace: storageNamespace, slot: slotId, leaseId: leaseId, generation: generation)
      return nil
    }
  }
  @objc func cancelInteraction(_ storageNamespace: String, slotId: String, accessibility: String,
                              resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    run(resolve, reject) { vault in
      guard let policy = VaultAccessibility(rawValue: accessibility) else { throw SessionVaultError.invalidInput }
      let result = try vault.cancelInteraction(namespace: storageNamespace, slot: slotId, accessibility: policy)
      return ["sessionJSON": result.sessionJSON as Any? ?? NSNull(), "cancelled": result.cancelled]
    }
  }
}
